### PR TITLE
Remove Data.Blob.File in favor of Analysis.File and use semantic-source’s Language type.

### PR DESCRIPTION
**Pull request by patrickt**
I apologize for the size of this patch—it is inherently long, and the change in our stylish-haskell formatting does not help. I recommend you disable whitespace changes.

### CODE REVIEW

File: bench/Tagging.hs [hunk 1]
```diff
@@ -66,7 +66,7 @@ parseSymbolsFilePath ::
   => PerLanguageModes
   -> Path.RelFile
   -> m ParseTreeSymbolResponse
-parseSymbolsFilePath languageModes path = readBlob (fileForTypedPath path) >>= runReader languageModes . parseSymbols . pure @[]
+parseSymbolsFilePath languageModes path = readBlob (File.fromPath path) >>= runReader languageModes . parseSymbols . pure @[]
```
Comment (patrickt) [inline bench/Tagging.hs]: Everything is with a typed API now.

File: semantic-analysis/src/Analysis/Language.hs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+extensionsForLanguage :: Language -> [String]
+extensionsForLanguage language = T.unpack <$> maybe mempty Lingo.languageExtensions (Map.lookup (languageToText language) Lingo.languages)
+
+forPath :: Path.PartClass.AbsRel ar => Path.File ar -> Language
+forPath path =
+  let spurious lang = lang `elem` [ "Hack" -- .php files
+                                  , "GCC Machine Description" -- .md files
+                                  , "XML" -- .tsx files
+                                  ]
+      allResults = Lingo.languageName <$> Lingo.languagesForPath (Path.toString path)
```
Comment (patrickt) [inline semantic-analysis/src/Analysis/Language.hs]: Could avoid a `toString` here if we pushed it into `lingo`, c.f. https://github.com/tclem/lingo-haskell/issues/6

File: semantic-ast/app/Main.hs [hunk 3]
```diff
@@ -19,9 +19,9 @@ import Data.ByteString.Lazy.Char8 (putStrLn)
 import Data.Aeson.Encode.Pretty (encodePretty)
 
 data SemanticAST = SemanticAST
-  { format :: Format
-  , noColor  :: Bool
-  , source :: Either [FilePath] String
+  { _format :: Format
+  , _noColor  :: Bool
+  , _source :: Either [FilePath] String
```
Comment (patrickt) [inline semantic-ast/app/Main.hs]: These were spawning some warnings with `cabal build all`.

File: src/Data/Blob.hs [hunk 4]
```diff
@@ -31,69 +31,52 @@ module Data.Blob
...
-  , fileLanguage :: Language
-  } deriving (Show, Eq)
-
--- | Prefer 'fileForTypedPath' if at all possible.
-fileForPath :: FilePath  -> File
-fileForPath p = File p (languageForFilePath p)
-
-fileForTypedPath :: Path.PartClass.AbsRel ar => Path.File ar -> File
-fileForTypedPath = fileForPath . Path.toString
+type File = Analysis.File.File Language
```
Comment (patrickt) [inline src/Data/Blob.hs]: This is a beautiful hack for compatibility purposes—in other projects, we store terms inside files, but because we do type gyrations elsewhere, we can replace the old `File` with the general `File` specialized to hold an associated `Language`.

File: src/Data/Blob.hs [hunk 5]
```diff
@@ -31,69 +31,52 @@ module Data.Blob
...
 blobLanguage :: Blob -> Language
-blobLanguage = fileLanguage . blobFile
+blobLanguage = Analysis.File.fileBody . blobFile
 
 blobPath :: Blob -> FilePath
-blobPath = filePath . blobFile
-
-makeBlob :: Source -> FilePath -> Language -> Text -> Blob
-makeBlob s p l = Blob s (File p l)
-{-# INLINE makeBlob #-}
```
Comment (patrickt) [inline src/Data/Blob.hs]: Took the chance to nuke this while I was in the neighborhood.

File: src/Data/Blob.hs [hunk 6]
```diff
@@ -31,69 +31,52 @@ module Data.Blob
...
-
-fileForTypedPath :: Path.PartClass.AbsRel ar => Path.File ar -> File
-fileForTypedPath = fileForPath . Path.toString
+type File = Analysis.File.File Language
 
 -- | The source, path information, and language of a file read from disk.
 data Blob = Blob
   { blobSource :: Source -- ^ The UTF-8 encoded source text of the blob.
   , blobFile   :: File   -- ^ Path/language information for this blob.
-  , blobOid    :: Text   -- ^ Git OID for this blob, mempty if blob is not from a git db.
```
Comment (patrickt) [inline src/Data/Blob.hs]: We no longer read items directly from Git, so the `oid` field was always empty. I’ve removed it.

File: src/Data/Language.hs [hunk 7]
```diff
@@ -111,53 +20,15 @@ supportedExts = foldr append mempty supportedLanguages
...
-  , goMode     :: LanguageMode
+  { pythonMode     :: LanguageMode
+  , rubyMode       :: LanguageMode
+  , goMode         :: LanguageMode
   , typescriptMode :: LanguageMode
-  , tsxMode :: LanguageMode
+  , tsxMode        :: LanguageMode
   , javascriptMode :: LanguageMode
-  , jsxMode :: LanguageMode
+  , jsxMode        :: LanguageMode
```
Comment (patrickt) [inline src/Data/Language.hs]: The `LanguageMode` stuff still lives inside `Data.Language` because it’s a `semantic`-specific concern.

File: src/Semantic/Api/Bridge.hs [hunk 8]
```diff
@@ -71,8 +78,18 @@ instance APIBridge T.Text Data.Language where
...
-    blobToApiBlob b = defMessage & P.content .~ toText (Data.blobSource b) & P.path .~ T.pack (Data.blobPath b) & P.language .~ (bridging # Data.blobLanguage b)
-    apiBlobToBlob blob = Data.makeBlob (fromText (blob^.content)) (T.unpack (blob^.path)) (blob^.(language . bridging)) mempty
+    blobToApiBlob b
+      = defMessage
+      & P.content .~ Source.toText (Data.blobSource b)
+      & P.path .~ T.pack (Data.blobPath b)
+      & P.language .~ (bridging # Data.blobLanguage b)
+    apiBlobToBlob blob =
+      let src = blob^.content.to Source.fromText
+          pth = fromRight (Path.toAbsRel Path.emptyFile) (blob^.path._Text.to Path.parse)
```
Comment (patrickt) [inline src/Semantic/Api/Bridge.hs]: This gyration is to handle the degenerate case of if we get an invalid path over the wire. This could conceivably violate the `Iso` laws but I truly do not find this important.

File: src/Data/Project.hs [hunk 9]
```diff
@@ -56,5 +57,5 @@ readProjectFromPaths maybeRoot path lang excludeDirs = do
   blobs <- liftIO $ traverse (readBlobFromFile' . toFile) paths
   pure $ Project (Path.toString rootDir) blobs lang (fmap Path.toString excludeDirs)
```
Comment (patrickt) [inline src/Data/Project.hs]: `Project` also needs a dose of `pathtype`, but I’ve chosen to do that in a separate PR, both because this patch is big enough already and because I think an OSS person wants to help out with that issue, which is fine by me, as I’ve written this patch like five times and I’m tired of it. 😇

File: src/Semantic/Task/Files.hs [hunk 10]
```diff
@@ -92,14 +101,9 @@ data FilesArg
   | FilesFromPaths [File]
 
 -- | A task which reads a list of 'Blob's from a 'Handle' or a list of 'FilePath's optionally paired with 'Language's.
-readBlobs :: (Has Files sig m, MonadIO m) => FilesArg -> m [Blob]
+readBlobs :: Has Files sig m => FilesArg -> m [Blob]
 readBlobs (FilesFromHandle handle) = send (Read (FromHandle handle) pure)
-readBlobs (FilesFromPaths [path]) = do
-  isDir <- isDirectory (filePath path)
```
Comment (patrickt) [inline src/Semantic/Task/Files.hs]: This cleans up some weird semantics where we were letting directory paths through and doing specious try-to-do-the-right-thing gyrations. It’s better to just reject special cases like this anyway, since our internal systems no longer hit this code path.

File: src/Semantic/Util.hs [hunk 11]
```diff
@@ -86,6 +93,9 @@ parseFile, parseFileQuiet :: Parser term -> FilePath -> IO term
 parseFile      parser = runTask'     . (parse parser <=< readBlob . fileForPath)
 parseFileQuiet parser = runTaskQuiet . (parse parser <=< readBlob . fileForPath)
 
+fileForPath :: FilePath -> File Language.Language
```
Comment (patrickt) [inline src/Semantic/Util.hs]: This is now internal API used only for REPL-level experiments, which is good enough for me.

File: semantic-analysis/src/Analysis/File.hs [hunk 12]
```diff
@@ -19,3 +24,10 @@ data File a = File
 fromBody :: HasCallStack => a -> File a
 fromBody body = File (Path.absRel (srcLocFile srcLoc)) (spanFromSrcLoc srcLoc) body where
   srcLoc = snd (fromJust (listToMaybe (getCallStack callStack)))
+
+-- | The language of the provided file, as inferred by 'Language.forPath'.
+fileLanguage :: File a -> Language
+fileLanguage = Language.forPath . filePath
+
+fromPath :: Path.PartClass.AbsRel ar => Path.File ar -> File Language
+fromPath p = File (Path.toAbsRel p) lowerBound (Language.forPath p)
```
Comment (patrickt) [inline semantic-analysis/src/Analysis/File.hs]: This used to be called `fileFromTypedPath`. Since `FilePath` is now no longer part of the API, we no longer need to differentiate, and can give it a simpler name.

File: src/Data/Blob.hs [hunk 13]
```diff
@@ -31,69 +27,52 @@ module Data.Blob
...
 import qualified System.FilePath as FP
 import qualified System.Path as Path
 import qualified System.Path.PartClass as Path.PartClass
 
--- | A 'FilePath' paired with its corresponding 'Language'.
--- Unpacked to have the same size overhead as (FilePath, Language).
-data File = File
-  { filePath     :: FilePath
-  , fileLanguage :: Language
-  } deriving (Show, Eq)
```
Comment (patrickt) [inline src/Data/Blob.hs]: The `File` type in `semantic-analysis` does not hold a `Language` member. However, it has an extra field, used in other places to hold parsed terms; we can stick a `Language` in there, thereby yielding a type equivalent in power to the old `File`.

File: src/Data/Language.hs [hunk 14]
```diff
@@ -1,107 +1,19 @@
...
-  let spurious lang = lang `elem` [ "Hack" -- .php files
-                                  , "GCC Machine Description" -- .md files
-                                  , "XML" -- .tsx files
-                                  ]
-      allResults = Lingo.languageName <$> Lingo.languagesForPath path
-  in case filter (not . spurious) allResults of
-    [result] -> textToLanguage result
-    _        -> Unknown
+codeNavLanguages :: [Language]
+codeNavLanguages = [Go, Java, Ruby, Python, JavaScript, TypeScript, PHP]
```
Comment (patrickt) [inline src/Data/Language.hs]: This still lives at the `semantic:lib` level, since it’s application-specific.

File: semantic-analysis/src/Analysis/Language.hs [hunk 15]
```diff
@@ -0,0 +1,136 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+module Analysis.Language
```
Comment (patrickt) [inline semantic-analysis/src/Analysis/Language.hs]: @tclem pointed out that this makes diffing, parsing, etc. have a dependency on `semantic-analysis` when we could get away with having it inside `semantic` alone. I could go either way on this; thoughts?

Comment (robrix) [reply semantic-analysis/src/Analysis/Language.hs]: IMO this file belongs in `semantic-source`, altho that is a bit more work at the moment due to us getting that package from hackage. (cf #423)

Comment (patrickt) [reply semantic-analysis/src/Analysis/Language.hs]: I’d wait until #423 to fix this, yeah.

File: semantic-analysis/src/Analysis/File.hs (hunk 12 above)
Comment (robrix) [reply semantic-analysis/src/Analysis/File.hs]: Excellent!

File: semantic-analysis/src/Analysis/Language.hs (hunk 15 above)
Comment (robrix) [reply semantic-analysis/src/Analysis/Language.hs]: I don’t think we should wait; the churn this introduces for the other packages makes it worth the extra effort for the `semantic-source` release, IMO.

If it helps, we could potentially roll this in with @aymannadeem’s intended changes to `semantic-source` to only do a single release; I’ll leave that to the two of you to suss out, tho.

Comment (robrix) [reply semantic-analysis/src/Analysis/Language.hs]: (Note to self: I’m also wondering if `Analysis.File` doesn’t belong in `semantic-source`, but that is orthogonal and should not hold up this PR.)

File: src/Data/Blob.hs (hunk 13 above)
Comment (robrix) [reply src/Data/Blob.hs]: We might consider it worth adding a language field to `Analysis.File.File`, but I’ll reserve judgement till we see how it feels.

File: src/Data/Blob.hs (hunk 6 above)
Comment (robrix) [reply src/Data/Blob.hs]: Nice 🤘

File: src/Data/Blob.hs [hunk 16]
```diff
@@ -31,69 +27,51 @@ module Data.Blob
...
-  { blobSource :: Source -- ^ The UTF-8 encoded source text of the blob.
-  , blobFile   :: File   -- ^ Path/language information for this blob.
-  , blobOid    :: Text   -- ^ Git OID for this blob, mempty if blob is not from a git db.
+  { blobSource :: Source         -- ^ The UTF-8 encoded source text of the blob.
+  , blobFile   :: File Language  -- ^ Path/language information for this blob.
   } deriving (Show, Eq)
 
 blobLanguage :: Blob -> Language
-blobLanguage = fileLanguage . blobFile
+blobLanguage = Analysis.File.fileBody . blobFile
```
Comment (robrix) [inline src/Data/Blob.hs]: Is `fileBody` ambiguous? Do we need the qualifier?

File: src/Data/Blob.hs [hunk 17]
```diff
@@ -112,7 +90,7 @@ moduleForBlob :: Maybe FilePath -- ^ The root directory relative to which the mo
               -> Module term    -- ^ A 'Module' named appropriate for the 'Blob', holding the @term@, and constructed relative to the root 'FilePath', if any.
 moduleForBlob rootDir b = Module info
   where root = fromMaybe (FP.takeDirectory (blobPath b)) rootDir
-        info = ModuleInfo (FP.makeRelative root (blobPath b)) (languageToText (blobLanguage b)) (blobOid b)
+        info = ModuleInfo (FP.makeRelative root (blobPath b)) (languageToText (blobLanguage b)) mempty
```
Comment (robrix) [inline src/Data/Blob.hs]: Can we remove the field from `ModuleInfo` as well or is that going to be a hassle?

File: src/Data/Blob.hs [hunk 18]
```diff
@@ -31,69 +27,51 @@ module Data.Blob
...
-
-inferringLanguage :: Source -> FilePath -> Language -> Blob
-inferringLanguage src pth lang
-  | knownLanguage lang = makeBlob src pth lang mempty
-  | otherwise = makeBlob src pth (languageForFilePath pth) mempty
+-- | Create a Blob from a provided path, language, and UTF-8 source.
+-- The resulting Blob's span is taken from the 'totalSpan' of the source.
+fromSource :: Path.PartClass.AbsRel ar => Path.File ar -> Language -> Source -> Blob
+fromSource filepath language source
+  = Blob source (Analysis.File.File (Path.toAbsRel filepath) (totalSpan source) language)
```
Comment (robrix) [inline src/Data/Blob.hs]: I appreciate the `fromBlah` naming scheme for smart constructors, thank you.

File: src/Data/Project.hs [hunk 19]
```diff
@@ -56,5 +57,5 @@ readProjectFromPaths maybeRoot path lang excludeDirs = do
...
   where
-    toFile path = File (Path.toString path) lang
+    toFile path = Analysis.File.File path lowerBound lang
```
Comment (robrix) [inline src/Data/Project.hs]: I think this doesn’t need to be qualified?

File: src/Semantic/Api/Bridge.hs (hunk 8 above)
Comment (robrix) [reply src/Semantic/Api/Bridge.hs]: Yeah, that seems reasonable 👍

File: src/Semantic/Api/Symbols.hs [hunk 20]
```diff
@@ -78,15 +88,15 @@ parseSymbols blobs = do
           & P.language .~ (bridging # blobLanguage')
           & P.symbols .~ mempty
           & P.errors .~ [defMessage & P.error .~ pack e]
-          & P.blobOid .~ blobOid
+          & P.blobOid .~ mempty
```
Comment (robrix) [inline src/Semantic/Api/Symbols.hs]: This is interesting.

1. What would be involved in updating the wire types to not include this field?

2. The wire types come with defaults for each field; does that make this equivalent to omitting the field?

3. If not, is the behaviour on the other end equivalent between this and omitting the field?

File: src/Semantic/CLI.hs [hunk 21]
```diff
@@ -1,10 +1,12 @@
-{-# LANGUAGE ApplicativeDo, FlexibleContexts #-}
+{-# LANGUAGE ApplicativeDo #-}
+{-# LANGUAGE FlexibleContexts #-}
 module Semantic.CLI (main) where
 
+import qualified Analysis.File as File
```
Comment (robrix) [inline src/Semantic/CLI.hs]: Pretty much convinced `Analysis.File` belongs in `semantic-source`, but again, that does not need to block this PR.

File: src/Semantic/Task/Files.hs (hunk 10 above)
Comment (robrix) [reply src/Semantic/Task/Files.hs]: I wonder if we should have a `-r` flag for recursive-mode to recover the other behaviour (with a separate code path, natch).

File: src/Semantic/Api/Symbols.hs (hunk 20 above)
Comment (patrickt) [reply src/Semantic/Api/Symbols.hs]: 1. We would need to track down all consumers of this API and make sure that they’re not using the `blobOid` field (which I don’t think they are).
2. Yes, the empty string is the default value for a `Text` field, so the behavior is preserved.
3. see above

Comment (robrix) [reply src/Semantic/Api/Symbols.hs]: Given that, should we omit the fields?

Comment (patrickt) [reply src/Semantic/Api/Symbols.hs]: I wouldn’t object, but I would do it in a separate PR, since that touches semanticd functionality as well.

Comment (robrix) [reply src/Semantic/Api/Symbols.hs]: Oh, I meant omit _setting_ the fields.

Comment (patrickt) [reply src/Semantic/Api/Symbols.hs]: Oh, I see. Yeah, roger that.

File: src/Data/Language.hs (hunk 14 above)
Comment (robrix) [reply src/Data/Language.hs]: Yeah that scans 👍
